[CS] Unify ReturnStmt handling

hamishknight opened the pull request:
Unify the implementation between single-expression and multi-statement closures. Because we're now storing a contextual type for single expression closure returns, update a code completion test to bring its behavior inline with the multi-statement case.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

xedin commented on lib/Sema/CSSyntacticElement.cpp:
```diff
@@ -1281,9 +1255,18 @@ class SyntacticElementConstraintGenerator
+
+    // Single-expression closures need their contextual type locator anchored
+    // on the closure itself.
+    if (context.isSingleExpressionClosure(cs) && returnStmt->hasResult()) {
```
Can this be a part of `getContextualResultInfo()` so we don't have to mutate it post factum?

hamishknight replied:
Good idea! How's this?

hamishknight commented:
@swift-ci please test

hamishknight commented:
Source compat run passed except for an issue that looks unrelated

xedin commented on lib/Sema/CSSyntacticElement.cpp:
```diff
@@ -1337,8 +1311,19 @@ class SyntacticElementConstraintGenerator
+      ConstraintLocator *loc = nullptr;
+      if (context.isSingleExpressionClosure(cs) && returnStmt->hasResult()) {
+        loc = cs.getConstraintLocator(
+            context.getAsAbstractClosureExpr().get(),
```
Nit: you can use `closure` here instead of re-fetching it

hamishknight replied:
Ack, good catch

xedin reviewed (approved):
Looks great!

hamishknight committed:
[CS] Remove `isInputExpression` parameter

This wasn't consistently used, and consequently
could result in some expressions getting their
parents invalidated. Instead, replace it with a
query to make sure we don't try and add an
expression we've already computed the parent info
for.

hamishknight committed:
[CS] Unify ReturnStmt handling

Unify the implementation between single-expression
and multi-statement closures. Because we're now
storing a contextual type for single expression
closure returns, update a code completion test to
bring its behavior inline with the multi-statement
case.

hamishknight commented:
@swift-ci please test

ahoppen reviewed (commented):
Change in code completion behavior looks reasonable to me.
